### PR TITLE
update dependencies and enable codecov

### PR DESCRIPTION
**Pull request by bjlittle**
This pull-request:

- updates the optional developer dependencies to include `pre-commit`
- updates the optional test dependencies to include `nox`, `codecov`, `pytest-cov`
- enables `codecov` test coverage reporting for the latest supported distribution of python i.e., `py312`
- rationalises the PyPI requirements

### CODE REVIEW

**Comment by bjlittle** [comment]
ping @tkknight :wink:

File: .github/workflows/ci-tests.yml [hunk 1]
```diff
@@ -123,14 +123,16 @@ jobs:
...
+          nox --session ${{ matrix.session }} -- ${{ matrix.coverage }} --test-data-dir ${HOME}/iris-test-data/test_data
 
-      - name: "Upload coverage report"
+      - name: "upload coverage report"
+        if: ${{ matrix.coverage }}
         uses: codecov/codecov-action@v4
-        if: ${{ matrix.coverage }}
\ No newline at end of file
+        with:
+          token: ${{ secrets.CODECOV_TOKEN }}
```
Comment (bjlittle) [inline .github/workflows/ci-tests.yml:138]: See https://github.com/codecov/codecov-action for more details

**Review by tkknight — APPROVED** [review]

**Comment by bjlittle** [comment]
@tkknight Awesome! Thanks 😄

**Review by pp-mo — APPROVED** [review]
After some thought, I do think I agree with all this !

**Pull request merged**